### ISSUE TITLE
I need some sort of iteration for tokenizers

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/patterns#2]
Currently I do something like this in a regex based parser:

```js
    parseParameters: function (string, pointer){
        var parameterLexer = /\w+|\)/g;
        parameterLexer.lastIndex = pointer;
        var parameters = [];
        while (true){
            var match = parameterLexer.exec(string);
            var token = match[0];
            if (token === closingParenthesis)
                break;
            parameters.push(token);
        }
        return parameters;
    }
```

**Comment by inf3rno** [comment]
I want to generalize somehow this Lexer. Actually only the `token === closingParenthesis` makes it a Lexer, otherwise it would be a Tokenizer...

**Comment by inf3rno** [comment]
I am not sure whether the lib should contain those terms, it would be a lot better to have some sort of iterator I could use in a while loop when I want to use the patterns in a tokenizer.

```js
    parseParameters: function (string, fromIndex){
        var parameterTokenizer = new PatternIterator("\\w+|\\)");
        var parameters = [];
        parameterTokenizer.setPointer(fromIndex);
        while (parameterTokenizer.findNext(string)){
            if (parameterTokenizer.match === closingParenthesis)
                break;
            parameters.push(parameterTokenizer.match);
        }
        return parameters;
    }
```

**Comment by inf3rno** [comment]
We could build a Lexer on top of this:

    parseParameters: function (string, fromIndex){
        var parameterLexer = new ParameterLexer();
        var parameters = [];
        parameterLexer.setPointer(fromIndex);
        var parameter;
        while (parameter = parameterLexer.nextParameter(string)) // param names cannot be falsy
            parameters.push(parameter);
        return parameters;
    }

**Comment by inf3rno** [comment]
How would the Lexer look like?

```js
function ParameterLexer(){
    this.tokenizer = new PatternIterator("\\w+|\\)");
}
ParameterLexer.prototype = {
    setPointer: function (pointer){
        this.tokenizer.setPointer(pointer);
    },
    findNext: function (string){
        if (!this.tokenizer.findNext(string))
            return false;
        if (this.tokenizer.match === ")")
            return false;
        return this.tokenizer.match;
    }
};

```

**Comment by inf3rno** [comment]
So it does the lexical analysis right after finding the next token with the tokenizer. It makes breaking the loop a lot simpler. It is not necessary to return the matching string from the tokenizer, we can attach metadata to it, e.g. we could return `new Token.Parameter(name)` instead of the string. I don't think the parameter names can be falsy in this case, so we can return the string too...

**Comment by inf3rno** [comment]
I think we need a `getPointer()` function too, because we used to use nested loops by languages with nested structures. Let's check the other tokenizer:

```js
    parseFunction(string, pointer){
        var functionLexer = /\(/g;
        functionLexer.lastIndex = pointer;
        functionLexer.exec(string);
        return this.parseParameters(string, functionLexer.lastIndex);
},
```

**Comment by inf3rno** [comment]
What I hate here that we use a function tokenizer for the param list opening separator and the closing separator. The ideal would be using the same tokenizer for both. Something like `new ParamListLexer(new ParamLexer())`.

**Comment by inf3rno** [comment]
If we want to separate those 2 patterns, then we need to race them somehow or import the closing automatically from the high level lexer.

**Comment by inf3rno** [comment]
Racing multiple regex patterns would be slow and impractical. Importing the closing token pattern would be the solution we need I think.

**Comment by inf3rno** [comment]
We should split up the code differently:
```js
	var functionTokenizer = new patterns.PatternIterator("\\(");
	functionTokenizer.setPointer(pointer);
	functionTokenizer.findNext(string);
	
	var parameterTokenizer = new PatternIterator("\\w+|\\)");
	var parameters = [];
	parameterTokenizer.setPointer(functionTokenizer.getPointer());
	while (parameterTokenizer.findNext(string)){
		if (parameterTokenizer.match === ")")
			break;
		parameters.push(parameterTokenizer.match);
	}
```

**Comment by inf3rno** [comment]
First we need to identify the token by capturing group, not by applying the subpattern again. So the `parameterTokenizer.match === ")"` is the wrong approach. We need to name the groups. We need named capturing group support from the pattern lib. I am not sure whether I need a separate Tokenizer class. I might merge it with the PatternIterator. But instead of the `"\\w+|\\)"` format we need to use a map for configuration: `{parameter: "\\w", closingParameters: "\\)"}`. The order of the keys stores the priority. Afaik. that works in IE8+, but I am not sure. In some of the old browsers the enumeration does not have a clear order, so we need to make the assumption that the environment is ok. Another approach is the array approach: `[{token: "parameter", pattern: "\\w"}, {token: "closingParameters", pattern: "\\)"}]`. The shortest we can do with the array approach is the following: `{parameter: "\\w"}, {closingParameters: "\\)"}` and we send it in the argument list of the constructor. Another possible solution to add the tokens one by one with a method: `tokenizer.addToken("parameter", "\\w")`.

**Comment by inf3rno** [comment]
I'd prefer the push, because that contains the implicit knowledge that the token is pushed to the end of the tokens array. What if we make that explicit?

**Comment by inf3rno** [comment]
So the tokenizer merges these token patters with `|`. We could have a rule, that the last given token breaks the tokenizer loop, but I'd rather avoid such implicit rules. A lot better way would be to add callbacks for the different tokens. Instead of naming the tokens we could name the callbacks or the callback parameters.

```js
var parameters = [];
var functionTokenizer = new patterns.Tokenizer();
functionTokenizer.addToken("\\(", function(match){
	var parameterTokenizer = new patterns.Tokenizer();
	parameterTokenizer.addToken("\\w+", function (match){
		parameters.push(match.substring);
	});
	parameterTokenizer.addToken("\\)", function (match){
		parameterTokenizer.stop();
	});
	parameterTokenizer.process(string, match.pointer);
});
functionTokenizer.process(string, pointer);
```

**Comment by inf3rno** [comment]
We could use iterators: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators instead of this `tokenizer.done()`. Actually a generator would be nice, but I don't want to depend on ES6 features.

**Comment by inf3rno** [comment]
Actually in this case it is obvious that we want to continue with a sub-tokenizer from the pointer with the same string. And I think this is something general... I am not entirely sure though, the pointer can be different for example if the token pattern has capturing groups.

**Comment by inf3rno** [comment]
We could still keep the upper long approach if we don't want to automatically connect the two tokenizers because the pointer will be set differently. So this is just sugar syntax...

**Comment by inf3rno** [comment]
It would be nice to pass somehow the `parameters` array as an argument, so we could do:
```js
return patterns.tokenizer(...).process(string, pointer);
```
But I don't think that is possible without a lot of additional code.

Another issue here that I want classes and reusable code. Reusing the generated regex patterns is not a real problem here, because they are cached, but still it would be nice to compile the pattern only once and reuse the same object in all of the cases. Which means we should do `this.tokenizer.process(string, pointer)`. Which means we definitely need to pass param and return something from the loop...

**Comment by inf3rno** [comment]
What if we support nested tokenizers and configure with a nested data structure?
```js
var parameterTokenizer = patterns.tokenizer(
	["\\(", "\\)"], [
		function (){
			this.parameters = [];
			this.run();
			return this.parameters;
		},
		"\\w+", function (parameter){
			this.parameters.push(parameter);
		}
	];
);
return parameterTokenizer.process(string, pointer);
```

**Comment by inf3rno** [comment]
It's ok, but it is far from being elegant.

```js
return patterns.tokenizer("\\(", [
	function middleware(next, string, pointer){
		this.parameters = [];
		next(string, pointer);
		return this.parameters;
	},
	"\\w+", function (parameter){
		this.parameters.push(parameter);
	}
], "\\)").process(string, pointer);
```

**Comment by inf3rno** [comment]
Another thought that the classical is the flat approach with an iterator:

```js
var tokenizer = patterns.tokenizer({
	parameters: ["\\(", {parameter: "\\w+"}, "\\)"]
}

let parameters;
for ({token, substring} of tokenizer.process(string, pointer)) {
	switch (token){
		case "parameters":
			parameters = [];
			break;
		case "parameter":
			parameters.push(substring);
			break;
	}
}
return parameters;
```

The only problem here that I don't know when we enter or exit from a level by nested tokenizers...

**Comment by inf3rno** [comment]
```js
var tokenizer = patterns.tokenizer({
	parameters: ["\\(", {parameter: "\\w+"}, "\\)"]
}

var parameters;
for ({token, substring, entering} of tokenizer.process(string, pointer))
	if (token === "parameters" && entering)
		parameters = [];
	else if (token === "parameter")
		parameters.push(substring);
return parameters;
```

**Comment by inf3rno** [comment]
The recursive version:

```js
var tokenizer = patterns.tokenizer({parameters: ["\\(", "parameters", {parameter: "\\w+"}, "\\)"]});
var levels = [null];
for ({token, substring, entering} of tokenizer.process(string, pointer))
	if (token === "parameters")
		if (entering)
			levels.push([]);
		else {
			levels[levels.length-2].push(levels[levels.length-1]);
			--levels.length;
		}
	else if (token === "parameter")
		levels[levels.length-1].push(substring);
return levels[0];
```

Ofc. I understand what I am doing, but it is cryptic asf.

**Comment by inf3rno** [comment]
If we want to hide all that leveling stuff we need to use function calls and let the stack handle it.

```js
var tokenizer = patterns.tokenizer({parameters: ["\\(", "parameters", {parameter: "\\w+"}, "\\)"]});
var ast = tokenizer.process(string, pointer, {
	parameters: function ({parent}){
		if (parent instanceof Array)
			parent.push([]);
		else
			parent.parameters = [];
	},
	parameter: function ({parent, substring}){
		parent.push(substring);
	}
});
return ast.parameters;

```

**Comment by inf3rno** [comment]
First: we should not call the process() and pass all that stuff directly. Actually we should not pass more than the `string` and `pointer` to the tokenizer. It's not the tokenizer's job to call all that stuff. That is the parser's job.
Second: the tokenizer is a lexer now, because it identifies tokens and adds metadata like the token type.

**Comment by inf3rno** [comment]
```js
class FunctionParamatersParser extends patterns.Parser {
	constructor(){
		this.compile({
			parameters: [
				"\\(", 
					"parameters", 
					{parameter: "\\w+"}, 
				"\\)"
			]
		});
	},
	parameters: function ({parent}){
		if (parent instanceof Array)
			parent.push([]);
		else
			return [];
	},
	parameter: function ({parent, substring}){
		parent.push(substring);
	}
}

var parser = new FunctionParamatersParser();
return parser.parse(string, pointer);
```

**Comment by inf3rno** [comment]
Actually we have a graph here by simple recursive rules. The only problem that this approach is not good for context dependent processing. So for example if the `parameters` can be recursive up to 3 levels, then the language is context-sensitive and this graph approach does not work, because we need additional rules to describe parsing. If the language is not recursive, then we can use a tree to describe it.

**Comment by inf3rno** [comment]
Interesting, so context-sensitive languages cannot be described by such general graphs. The syntactic validity can depend on the context. On top of that by natural languages the semantics can depend on the context too. Actually semantical changes is not a problem here, the `if (parent instanceof Array)` is for that in the current code, because we call the same method for both of the "()" tokens. The same way we can add another if-else blocks to give different meanings to the same token.

**Comment by inf3rno** [comment]
Another interesting thing here, that 

					"parameters", 
					{parameter: "\\w+"}, 

Here we could call the `parameter` processor on the nested `parameters`. At least the `parent.push()` is not reused here. We need to reuse somehow. Which means we need to generalize the collection and the item somehow. The `parameters` describes a collection, while the `parameter` and the recursion describes an item of that collection.

**Comment by inf3rno** [comment]
Another issue here is naming, since we don't process every token, just the ones the pattern matches. This is not a problem in some cases, while in others it is. So we need to name the parser accordingly.

**Comment by inf3rno** [comment]
I think the `this.compile(graph)` should not be part of the parser. Let's assume we have a different lexer, which splits up sentences by whitespace. In that case we can still use pattern matching on the words (tokens) we get. We can use it similarly with a changing `token1|token2|token3|...` pattern, still the regex we will use will be something like `/^(?:(token1)|(token2)|(token3))$/g` instead of `/(token1)|(token2)|(token3)/g`. The processing is different too, we can use `/\s+/g` for finding token boundary and check the non-matching with the upper token pattern. We can even split the text by this pattern before starting to identify tokens, so obviously there are 2 type of regex based lexer algorithms here. That means we need a general Lexer interface and move the changing part behind that interface.

**Comment by inf3rno** [comment]
```js
class FunctionParamatersParser extends patterns.Parser {
	constructor(){
		this.lexer = new patterns.PatternMatchingLexer({
			parameters: [
				"\\(", 
					"parameters", 
					{parameter: "\\w+"}, 
				"\\)"
			]
		});
	},
	parameters: function ({parent}){
		if (parent instanceof Array)
			parent.push([]);
		else
			return [];
	},
	parameter: function ({parent, substring}){
		parent.push(substring);
	},
	nonMatching: function ({substring}){
	}
}
```

**Comment by inf3rno** [comment]
An alternative solution:

```js
var lexer = new patterns.PatternMatchingLexer({
	parameters: [
		"\\(", 
			"parameters", 
			{parameter: "\\w+"}, 
		"\\)"
	]
});

var parser = new patterns.Parser(lexer, {
	parameters: function ({parent}){
		if (parent instanceof Array)
			parent.push([]);
		else
			return [];
	},
	parameter: function ({parent, substring}){
		parent.push(substring);
	},
	nonMatching: function ({substring}){
	}
});

var parameters = parser.parse(string, pointer);
```

**Comment by inf3rno** [comment]
Reusing code is not as hard as I thought:

```js
var lexer = new patterns.PatternMatchingLexer({
	parameters: [
		"\\(", 
			"parameters", 
			{parameter: "\\w+"}, 
		"\\)"
	]
});

var parser = new patterns.Parser(lexer, {
	parameters: function ({parent}){
		if (parent instanceof Array)
			this.parameter({parent,substring:[]});
		else
			return [];
	},
	parameter: function ({parent, substring}){
		parent.push(substring);
	},
	nonMatching: function ({substring}){
	}
});

var parameters = parser.parse(string, pointer);
```

**Comment by inf3rno** [comment]
What we still need is generalizing AST building in a simple explicit way.

Doing this is not acceptable:
```js
	parameters: function ({parent}){
		if (parent instanceof Array)
			this.parameter({parent,substring:[]});
		else
			return [];
	},
 ```

We have no idea what return means in this context. Actually that is used for building the root node I think. But I worked with this 3 days ago and I totally forgot...

**Comment by inf3rno** [comment]
Another thought:
```js
var parameters = parser.parse(string, pointer);
```
I am not sure if we need the pointer here. We could slice the string instead and make it simpler. It would take a little bit more memory. I am not sure though. If we have a big string and we use multiple different lexers, then I think it make a difference.

**Comment by inf3rno** [comment]
Well another option is injecting the string and the pointer with the lexer and publish the results with events, but it would feel weird in this case. It would be a working solution by streams though. `raw -> lexer -> parser -[params]->`. Currently we need this for sync solutions only, so better to do it in a sync way.

**Comment by inf3rno** [comment]
I think a basic return value independent solution would look like this:

```js
class MyParser extends patterns.Parser {
	constructor(){
		super();
		this.lexer = new patterns.PatternMatchingLexer({
			parameters: [
				"\\(", 
					"parameters", 
					{parameter: "\\w+"}, 
				"\\)"
			]
		});
	}
	parameters(closing){
		if (closing) {
			--this.current.length;
			--this.level;
			return;
		}
		if (this.level > 0) {
			var branch = [];
			var parent = this.current[this.level];
			parent.push(branch);
			this.current.push(branch);
		}
		++this.level;
	}
	parameter(name){
		var parent = this.current[this.level];
		parent.push(name);
	},
	parse(string){
		this.tree = [];
		this.current = [this.tree];
		this.level = 0;
		super.parse(string);
		return this.tree;
	}
}

var parameters = parser.parse(string);
```

**Comment by inf3rno** [comment]
If we are building a tree, then this would work too:

```js
class MyParser extends patterns.Parser {
	constructor(){
		super();
		this.lexer = new patterns.PatternMatchingLexer({
			parameters: [
				"\\(", 
					"parameters", 
					{parameter: "\\w+"}, 
				"\\)"
			]
		});
	}
	start(next){
		return next();
	}
	parameters(next){
		var parameters = [];
		while (var parameter = next())
			parameters.push(parameter);
		return parameters;
	}
	parameter(parameter){
		return parameter;
	}
}

var parameters = new MyParser().parse(string);
```

**Comment by inf3rno** [comment]
I think the most complicated example is building a graph with multiple lexers. The multilexer is a viable solution when a block is on a different language.

**Comment by inf3rno** [comment]
I am pretty sure this is no good for context sensitive languages. I like this one better:

```js
class MyParser {
	parse(string){
		var lexer = new patterns.PatternMatchingLexer({
			parameters: ["\\(", "\\)"],
			parameter: "\\w+"
		});
		lexer.subject(string);
		return lexer.callback(lexer.query(lexer.tokens.parameters), this);
	}
	parameters(lexer){
		var parameters = [];
		var query = lexer.query(lexer.tokens.parameters, lexer.tokens.parameter);
		while (var parameter = lexer.callback(query, this))
			parameters.push(parameter);
		return parameters;
	}
	parameter(lexer){
		return lexer.matchingSubstring;
	}
}

var parameters = new MyParser().parse(string);
```

**Comment by inf3rno** [comment]
Reusing the lexer and the queries is an interesting task. Normally a lexer can handle only a single string. We could make pointers explicit and that way decouple it from the actual string. We could decouple it from the tokens too.

What should the lexer know?

It is maintaining the token tree. It knows which blocks are open and what to add to the regex patterns to close them. E.g. `...|\)` for closing the parameter list. It combines the patterns of the given tokens. The simplest way is doing `or(a,b,c) -> a|b|c`. We can do `and(a,b,c) -> abc` too. The current lexer does not really care about gaps (non-matching string). We can write a lexer which does, but I don't care currently. We should make this stable first. After that I will work on reusing lexer, queries, regex, etc...

**Comment by inf3rno** [comment]
What I really wanted is something really high level, when I hide that we are using regex in the low level code and we generate the regex patterns.

One way to move it to higher level, so the parser does not have to know about it:

```js
class MyParser {
	parse(lexer){
		return lexer.callback(lexer.query(lexer.tokens.parameters), this);
	}
	parameters(lexer){
		var parameters = [];
		var query = lexer.query(lexer.tokens.parameters, lexer.tokens.parameter);
		while (var parameter = lexer.callback(query, this))
			parameters.push(parameter);
		return parameters;
	}
	parameter(lexer){
		return lexer.matchingSubstring;
	}
}

var lexer = new PatternMatchingLexer({
	parameters: [, "\\)"],
	parameter: "\\w+"
});
lexer.subject(string);
var parameters = new MyParser().parse(lexer);
```

Or we can add regex as default engine and do something total sugar:

```js
{createLexer, charsets: {utf8: utf8}} = require("patterns");

class MyParser {
	parse(string){
		var lexer = createLexer({
			parameters: "()".split(""),
			parameter: utf8.word
		});
		lexer.subject(string);
		lexer.mapping(this);
		return lexer.callback(lexer.tokens.parameters);
	}
	parameters(lexer){
		var parameters = [];
		var parameter;
		var query = lexer.or(lexer.tokens.parameters, lexer.tokens.parameter);
		while (parameter = lexer.callback(query))
			parameters.push(parameter);
		return parameters;
	}
	parameter(lexer){
		return lexer.matchingSubstring;
	}
}


var parameters = new MyParser().parse(string);
```

**Comment by inf3rno** [comment]
I think as long as the usage of the lexer is something regex independent it does not really matter how the lexer was intantiated or how we injected the patterns. I am not sure if I can make a regex independent lexer syntax or if I want to. If should contains the queryStack that is certain. Another option to use a tokenizer instead and let the parser check the syntax. In that case we would do something like this:

```js
class MyParser {
	parse(string){
		var tokenizer = new patterns.RegexTokenizer({
			parameters: {
				open: "\\(",
				close: "\\)"
			},
			parameter: "\\w+"
		});
		tokenizer.subject(string);
		tokenizer.find(tokenizer.tokens.parameters.open);
		return this.parameters(tokenizer);
	}
	parameters(tokenizer){
		var parameters = [];
		var match;
		while (match = tokenizer.find(tokenizer.tokens.parameters.open, tokenizer.tokens.parameter, tokenizer.tokens.parameters.close))
			if (match.token === tokenizer.tokens.parameters.open)
				parameters.push(this.parameters(tokenizer));
			else if (match.token === tokenizer.tokens.parameter)
				parameters.push(this.parameter(match));
			else if (match.token === tokenizer.tokens.parameters.close)
				break;
		return parameters;
	}
	parameter(match){
		return match.substring;
	}
}
```

**Comment by inf3rno** [comment]
We can reduce the upper to a regex pattern matching helper. Something like:

```js
match = helper.find(string, fromIndex, "(\\()|(\\w)|(\\))", ["parameters.open", "parameter", "parameters.close"]);
match.index; // 0
match.name; // "parameters.open"
match.range; // [firstPosition, lastPosition]
match.groups; // {}
match.substring; // "("
```

**Comment by inf3rno** [comment]
I think this should be handled by #38 #34 #24 . If we are ready with those we can come back here and write parsers.

**Comment by inf3rno** [comment]
I renamed this, because this should be the v1.0 milestone. If this issue is done I am done with all of the 1.0 features.

**Comment by inf3rno** [comment]
As far as I understand having multiple patterns and iterators are enough to support tokenizers for context free grammar. In the long run we might be able to describe the same with recursion and without iterators. It would be fun to define an entire language in a single recursive pattern. I'll rename this to indicate that we are talking about context free grammar here.
